src/codegen/generators/pou_generator.rs

```diff
@@ -536,7 +563,57 @@ impl<'ink, 'cg> PouGenerator<'ink, 'cg> {
                         m.source_location.get_column(),
                     );
                 }
-                self.llvm.builder.build_store(ptr, ptr_value);
+
+                if matches!(m.argument_type, ArgumentType::ByVal(VariableType::Input))
+                    && type_info.is_aggregate()
+                {
+                    // a by-value aggregate type => we need to memcpy the data into the local variable
+                    let ty = if ty.is_array_type() {
+                        ty.into_array_type()
+                            .get_element_type()
+                            .ptr_type(AddressSpace::from(ADDRESS_SPACE_GENERIC))
+                    } else {
+                        ty.into_struct_type().ptr_type(AddressSpace::from(ADDRESS_SPACE_GENERIC))
+                    };
+                    let bitcast = self.llvm.builder.build_bitcast(ptr, ty, "bitcast").into_pointer_value();
+                    let (size, alignment) = if let DataTypeInformation::String { size, encoding } = type_info
+                    {
+                        // since passed string args might be larger than the local acceptor, we need to first memset the local variable to 0
```

Comment:
Can you elaborate on this? I still don't understand why the memset is needed here 😅 I initially thought the memset was required to avoid garbage values in the `alloca` call but that's not the case here?

Reply:
We don't copy the entire length of the locally allocated string, since the passed string might be larger than the acceptor:
```
FUNCTION foo
VAR_INPUT
	str: STRING[3];
END_VAR
END_FUNCTION

foo('longer than 3');
```
And to ensure the last grapheme is in fact a null-terminator, we memset the entire string to 0. An alternative would be to GEP into the last element and set it to 0 ourselves, but I'm not sure if that is worth it